Batched small updates

leafpetersen opened the pull request:
Several small updates to questions and clarifications that have been resolved in the issue tracker but not moved to the spec.

lrhn commented on resources/type-system/flow-analysis.md:
```diff
@@ -432,36 +435,22 @@ Definitions:
+    type `T` to a variable of type `S`, that is:
+    - If the demotion policy applies, treat the assignment as a demoting
+      assignment which demotes `x` to `T`.
+    - Otherwise, it is an error, unless `T` is dynamic, in which case it is an
```
`dynamic` is assignable to `S`, so it wouldn't reach here.

What type is the implicit downcast to? I'm assuming `S`.

leafpetersen replied:
True, and yes.

lrhn commented on resources/type-system/flow-analysis.md:
```diff
@@ -432,36 +435,22 @@ Definitions:
+    - If the demotion policy applies, treat the assignment as a demoting
+      assignment which demotes `x` to `T`.
+    - Otherwise, it is an error, unless `T` is dynamic, in which case it is an
+      implicit downcast.
```
Would it rather be:

> * If `T` is a subtype of `S`, then if `T` is a type of interest for `x`, `x` is promoted to `T`, 
>   and if it is not a type of interest, nothing further happens to the type of `x`. *(This includes `T` being `S`.)*
> * If the demotion policy applies, treat the assignment as a demoting assignment which demotes `x` to `T`.
> * Otherwise if `T` is `dynamic`, then perform an implicit downcast from `T` to `S` on the value of `E`.
> * Otherwise the assignment is a compile-time error.


(I assume that context-type based conversions like callable-object tear-off has happened during inference of `E`).

leafpetersen replied:
Rather than re-specify all of the promotion/demotion policies, I've just changed this to say "at this point, treat it like any other assignment of something of type `T` to a variable of type `S`."  I don't think we need any further text here do we?  The point is just that we use the current type of `x` for downwards inference, but still consider the assignment as a potential promotion/demotion point.

lrhn commented on resources/type-system/inference.md:
```diff
@@ -212,15 +216,22 @@ unlike normal fields, the initializer for a `late` field may reference `this`.
 Function literals which are inferred in an empty typing context (see below) are
 inferred using the declared type for all of their parameters.  If a parameter
-has no declared type, it is treated as if it was declared with type `dynamic`.
+has no declared type, it is treated as if it was declared with type `dynamic.
```
Missing a final \` after `` `dynamic``

leafpetersen replied:
Done.

leafpetersen commented on accepted/future-releases/nnbd/feature-specification.md:
```diff
@@ -1215,6 +1225,31 @@ all occurrences of `S*` in `T` shall be replaced with `S`.  As a result, legacy
+### Typedefs defined in legacy libraries used in opted-in libraries
+
+A typedef which is define in a legacy library and used in an opted-in library is
+treated defining a function type all of the components of which are treated as
```
Done.

leafpetersen commented on resources/type-system/flow-analysis.md:
```diff
@@ -432,36 +435,22 @@ Definitions:
+    context.  All reference to `x` within `E` are treated as having type `S` as
+    usual.
+  - Let `T` be the resulting inferred type of `E`.
+  - If `T` is assignable to `S` then no further work is required.
```
Simplified this to just say that the usual promotion/demotion/error rules for an assignment of an expression of type `T` to a variable of current type `S` apply.

leafpetersen commented:
Ok I addressed comments, landing this.  I'll follow up if more comments come in on the revised text.

cc @scheglov @johnniwinther for visibility.  I believe all of the changes here are things we already clarified in issues and are most likely already implemented.  I am adding some quick tests to verify.

leafpetersen commented:
[Tests](https://dart-review.googlesource.com/c/sdk/+/150160) out for review.

yzhong52 commented on accepted/future-releases/nnbd/feature-specification.md:
```diff
@@ -6,6 +6,11 @@ Status: Draft
 
 ## CHANGELOG
 
+2020..06.02
```
FYI: double dots here.

eernstg replied:
Fixed in 8c3bf5e2ba6b6f38dbf66fd7a39e2f84a3463eba.

johnniwinther commented on accepted/future-releases/nnbd/feature-specification.md:
```diff
@@ -1215,6 +1225,31 @@ all occurrences of `S*` in `T` shall be replaced with `S`.  As a result, legacy
 
+### Typedefs defined in legacy libraries used in opted-in libraries
+
+A typedef which is define in a legacy library and used in an opted-in library is
```
defined
